File: vault/frontend/src/services/speechChunks.ts
// Speaking San's reply a sentence at a time instead of all at once.
//
// Kokoro renders the WHOLE reply before returning a single byte, and synthesis time
// scales with length — measured on Everest at roughly 31 characters per second, so a
// 436-character answer is 12 seconds of silence before you hear anything. That silence
// sits on top of the model's own time, and it was the most irritating part of a voice
// call by some distance.
//
// Splitting on sentence boundaries lets the first clause start playing while the rest
// is still being made. Speech plays at roughly 12-13 characters per second and Kokoro
// synthesises at ~31, so once the first chunk is playing the pipeline stays comfortably
// ahead of the ear and the remaining joins are inaudible.
//
// The first chunk is deliberately much shorter than the rest: it is the only one whose
// synthesis time the listener actually waits through.

// A sentence over the first cap used to go out whole; on a measured call that was 229
// characters and 5.4s before the first word. The opening sentence is now cut at a clause
// boundary instead, matching the iPhone app.
const FIRST_MAX = 80;
const REST_MAX  = 220;   // ~7s to synthesise, ~17s to speak
const MIN_HEAD  = 25;    // never cut the opening so early it is a fragment, not a phrase

// Sentence-ish boundaries. Deliberately conservative: a split in the wrong place is
// audible as an odd pause, whereas a chunk that runs long only costs a little latency.
const BOUNDARY = /(?<=[.!?])\s+|\n+/;
const SPOKEN = /[\p{L}\p{N}]/u;
const CLAUSE_MARKS = ',;:—–';

// Cuts at the last clause mark (followed by a space) before `max`, else the last space.
// Null when there is no sensible cut, and the sentence then goes out whole.
function breakEarly(s: string, max: number): [string, string] | null {
  const last = Math.min(max, s.length) - 1;
  let cut = -1;
  for (let i = last; i >= MIN_HEAD; i--) {
    if (CLAUSE_MARKS.includes(s[i]) && s[i + 1] === ' ') { cut = i + 1; break; }
  }
  if (cut < 0) {
    for (let i = last; i >= MIN_HEAD; i--) {
      if (s[i] === ' ') { cut = i; break; }
    }
  }
  if (cut < 0) return null;
  const head = s.slice(0, cut).trim();
  const tail = s.slice(cut).trim();
  return head && tail ? [head, tail] : null;
}

export function splitForSpeech(text: string): string[] {
  const clean = text.trim();
  if (!clean) return [];

  // A "?!" or a stray "." after an emoji arrives as a piece of its own, and sent alone it
  // is a whole TTS request for no sound. Pieces with nothing to pronounce join the one
  // before them, or are dropped when there is none.
  const pieces: string[] = [];
  for (const p of clean.split(BOUNDARY).map(s => s.trim()).filter(Boolean)) {
    if (SPOKEN.test(p)) pieces.push(p);
    else if (pieces.length) pieces[pieces.length - 1] += ' ' + p;
  }

  // Only the opening chunk is worth cutting mid-sentence: it is the one being waited for.
  if (pieces.length && pieces[0].length > FIRST_MAX) {
    const cut = breakEarly(pieces[0], FIRST_MAX);
    if (cut) pieces.splice(0, 1, cut[0], cut[1]);
  }

  const chunks: string[] = [];
  let buf = '';

  for (const piece of pieces) {
    const max = chunks.length === 0 ? FIRST_MAX : REST_MAX;
    if (!buf) { buf = piece; }
    else if (buf.length + 1 + piece.length <= max) { buf += ' ' + piece; }
    else { chunks.push(buf); buf = piece; }

    // A single sentence longer than the cap still has to go out on its own — better a
    // long chunk than a split mid-clause, which sounds broken.
    if (buf.length >= max) { chunks.push(buf); buf = ''; }
  }
  if (buf) chunks.push(buf);
  return chunks;
}

export interface SpeakChunksOptions {
  // Turns one chunk of text into a playable object URL.
  synth: (text: string) => Promise<string>;
  // Called with each <audio> as it starts, so callers can hold it for barge-in.
  onAudio?: (audio: HTMLAudioElement) => void;
  // Checked between and during chunks; true stops the rest.
  shouldStop?: () => boolean;
  // Awaited while a chunk plays. Lets call mode watch the mic for interruptions;
  // resolves early if the caller wants to cut San off. Returns true to stop.
  awaitPlayback?: (audio: HTMLAudioElement) => Promise<boolean>;
}

export async function speakChunks(text: string, opts: SpeakChunksOptions): Promise<void> {
  const chunks = splitForSpeech(text);
  if (chunks.length === 0) return;

  // Synthesis of chunk N+1 is kicked off before chunk N finishes playing, which is what
  // removes the gaps. Only ONE is ever in flight ahead — running them all at once would
  // queue behind each other on the TTS server and delay the one actually needed next.
  let next: Promise<string> | null = opts.synth(chunks[0]);

  for (let i = 0; i < chunks.length; i++) {
    if (opts.shouldStop?.()) break;

    let url: string;
    try {
      url = await next!;
    } catch {
      // A failed chunk ends the utterance rather than skipping it — a reply missing its
      // middle sentence is worse than one that stops short.
      break;
    }

    next = i + 1 < chunks.length ? opts.synth(chunks[i + 1]) : null;

    const audio = new Audio(url);
    opts.onAudio?.(audio);
    try {
      await audio.play();
      const stopped = opts.awaitPlayback
        ? await opts.awaitPlayback(audio)
        : await new Promise<boolean>(resolve => {
            audio.onended = () => resolve(false);
            audio.onerror = () => resolve(false);
          });
      if (stopped) { URL.revokeObjectURL(url); break; }
    } finally {
      URL.revokeObjectURL(url);
    }
  }

  // Drop a prefetch nobody will play, so its object URL isn't leaked.
  next?.then(URL.revokeObjectURL).catch(() => {});
}

File: ios/MaayaCompanion/MaayaCompanion/Services/SpeechChunks.swift
import Foundation

// Splits a reply so San starts talking before the whole thing has been synthesised.
//
// Kokoro renders an entire reply before returning any audio, so a long answer meant
// several seconds of silence with nothing on screen either. The web client hit the same
// wall and solved it this way: speak the first sentence as soon as it exists, synthesise
// the next one while the current is playing. Measured there, a 251-character reply went
// from 8.1s to 2.5s before the first word.
//
// The first chunk is deliberately much shorter than the rest. It is the only one the
// listener actually waits for -- every later chunk is being made while the previous one
// plays, so its size costs nothing as long as synthesis keeps ahead of playback.
enum SpeechChunks {
    // Measured on a call: a 229-character opening sentence took 5.4s to synthesise,
    // because a sentence over the limit used to go out whole. The first chunk is now cut
    // at a clause boundary instead, so the wait before the first word stays short.
    static let firstMax = 80
    static let restMax  = 220

    // Never cut the first chunk so early that it is a fragment rather than a phrase.
    private static let minHead = 25

    private static let terminators: Set<Character> = [".", "!", "?", "\n"]
    private static let clauseMarks: Set<Character> = [",", ";", ":", "—", "–"]

    static func split(_ text: String) -> [String] {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        var queue = mergeFragments(sentences(of: trimmed))
        var chunks: [String] = []
        var current = ""
        var limit = firstMax
        var i = 0

        while i < queue.count {
            let sentence = queue[i]
            i += 1

            if current.isEmpty && sentence.count > limit {
                // Only the opening chunk is worth cutting mid-sentence: it is the one being
                // waited for. Later ones go out whole, because half a sentence read aloud
                // sounds broken and they are synthesised while earlier audio plays.
                if chunks.isEmpty, let (head, tail) = breakEarly(sentence, within: firstMax) {
                    chunks.append(head)
                    queue.insert(tail, at: i)
                } else {
                    chunks.append(sentence)
                }
                limit = restMax
                continue
            }
            if !current.isEmpty && current.count + sentence.count + 1 > limit {
                chunks.append(current)
                current = ""
                limit = restMax
                i -= 1                     // weigh this sentence again against the new limit
                continue
            }
            current = current.isEmpty ? sentence : current + " " + sentence
        }
        if !current.isEmpty { chunks.append(current) }
        return chunks
    }

    // Cuts at the last clause mark (followed by a space) before `max`, else the last space.
    // Nil when there is no sensible cut, and the sentence then goes out whole.
    private static func breakEarly(_ s: String, within max: Int) -> (String, String)? {
        let chars = Array(s)
        let last = min(max, chars.count) - 1
        guard last >= minHead else { return nil }

        var cut: Int?
        for idx in stride(from: last, through: minHead, by: -1)
        where clauseMarks.contains(chars[idx]) && idx + 1 < chars.count && chars[idx + 1] == " " {
            cut = idx + 1
            break
        }
        if cut == nil {
            for idx in stride(from: last, through: minHead, by: -1) where chars[idx] == " " {
                cut = idx
                break
            }
        }
        guard let c = cut else { return nil }

        let head = String(chars[..<c]).trimmingCharacters(in: .whitespaces)
        let tail = String(chars[c...]).trimmingCharacters(in: .whitespaces)
        guard !head.isEmpty, !tail.isEmpty else { return nil }
        return (head, tail)
    }

    // A "?!" or a stray "." after an emoji arrives as a sentence of its own. Sent alone it
    // was a whole TTS request -- measured at 5.7s, queued behind the real speech -- for no
    // sound at all. Pieces with nothing to pronounce join the previous piece, or are
    // dropped when there is none.
    private static func mergeFragments(_ pieces: [String]) -> [String] {
        var out: [String] = []
        for p in pieces {
            if hasSpokenContent(p) {
                out.append(p)
            } else if !out.isEmpty {
                out[out.count - 1] += p
            }
        }
        return out
    }

    private static func hasSpokenContent(_ s: String) -> Bool {
        s.unicodeScalars.contains { CharacterSet.alphanumerics.contains($0) }
    }

    // Keeps the terminator attached — the speech model uses it for prosody, and a
    // question read as a statement is immediately noticeable.
    private static func sentences(of text: String) -> [String] {
        var out: [String] = []
        var buf = ""
        for ch in text {
            buf.append(ch)
            if terminators.contains(ch) {
                let piece = buf.trimmingCharacters(in: .whitespacesAndNewlines)
                if !piece.isEmpty { out.append(piece) }
                buf = ""
            }
        }
        let tail = buf.trimmingCharacters(in: .whitespacesAndNewlines)
        if !tail.isEmpty { out.append(tail) }
        return out
    }
}
